[Storage] ListAccountSas and ListServiceSas

JasonYang-MSFT opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [x] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

JasonYang-MSFT committed:
Initialize version 2016-05-01

JasonYang-MSFT committed:
[Storage] Add listAccoutSas/listServiceSas operation

JasonYang-MSFT committed:
[Storage] updated returned type doc for ListAccountSas and List Service Sas

azurecla commented:
Hi __@JasonYang-MSFT__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (yabo). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

fearthecowboy commented:
Greetings @JasonYang-MSFT ,

Joahannes (@olydis), Deepak (@dsgouda) and I (@fearthecowboy) are the assigned reviewers for your swagger spec. (we're taking over for the great and powerful @amarzavery, who is really busy :D )

You can think of as your sherpas to guide your journey to acceptance thru this process :D

If you have any questions or need some help with our review, feel free to ask us anything.

olydis commented:
Hey @JasonYang-mft

You have clicked the checkbox in the post at the top that says: "Validation errors from the Linter extension for VS Code have all been fixed ..."

We've found a number of issues including the following:
- Typo: `canonicalizedResrouce` 
- couple of `description`s missing

Please recheck that you've used the Validation in VSCode and corrected all the errors before pushing the updates.

Once you've corrected the errors, please comment back and we can start with the PR review.

olydis reviewed (changes requested):
please address comments

fearthecowboy reviewed (changes requested):
.

robinsh commented:
In the description fields, can you please capitalize SAS, because technically it's an abbreviation or acronym.

JasonYang-MSFT committed:
[Storage] rename sas to SAS, add more descriptions.

JasonYang-MSFT commented:
Hi @olydis,
Thanks for your comments. I have addressed those issues. No more issues showed in VSCode with Linter extention.

JasonYang-MSFT commented:
Hi @robinsh 
I have changed "Sas", "sas" in descriptions to "SAS". Also add some missing descriptions. Can you take a look on that too?

fearthecowboy commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+            "required": true,
+            "schema": {
+              "$ref": "#/definitions/StorageAccountCheckNameAvailabilityParameters"
+            },
```
Given that the parameter object is only a couple fields, you may want to use the  "x-ms-client-flatten" extension here so that the properties of the parameters object are just promoted to parameters in the generated code. (see [https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-client-flatten](https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-client-flatten) )

robinsh replied:
This spec has only minor changes from the previous version. It may look like a new spec, but it's not. I don't understand why major changes would be needed.

JasonYang-MSFT replied:
Yes. Only two new features are added compared to previous version. Other features are same to previous version.
https://github.com/Azure/azure-rest-api-specs/pull/742/files#diff-5b8b9619ecdc1dedc0bfccf4e2662d8aR417
https://github.com/Azure/azure-rest-api-specs/pull/742/files#diff-5b8b9619ecdc1dedc0bfccf4e2662d8aR467

Once an api version is released(say 2016-01-01), feature team would like to working on new features to a new api version instead of adding things to a released version. That is also why we update the api version. I have an initial commit to copy swagger from 2016-01-01 to 2016-05-01, you can compare changes to that commit to see newly added changes.
https://github.com/Azure/azure-rest-api-specs/pull/742/commits/89e216ea837a23c5a77102d340c78d6563e98889

amarzavery replied:
@fearthecowboy - Your recommendation is already being implemented. It is a code generation setting on the command line (-FT 2).'
[generate.cmd in the .net sdk](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/src/ResourceManagement/Storage/Microsoft.Azure.Management.Storage/generate.cmd#L17)

call "%repoRoot%\tools\autorest.gen.cmd" %specFile% Microsoft.Azure.Management.Storage %autoRestVersion% %generateFolder% MICROSOFT_MIT **"-FT 2"**

fearthecowboy commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+        "description": "Asynchronously creates a new storage account with the specified parameters. If an account is already created and a subsequent create request is issued with different properties, the account properties will be updated. If an account is already created and a subsequent create or update request is issued with the exact same set of properties, the request will succeed.",
+        "parameters": [
+          {
+            "name": "resourceGroupName",
```
Common path parameters can be placed in the global parameters section and referenced.  

Just make sure you mark it with "x-ms-parameter-location" (see [https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-parameter-location](https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-parameter-location))

fearthecowboy replied:
resourceGroupName typically has the following constraints:

 "minLength": 3,
 "pattern": "[a-zA-Z0-9]+"

amarzavery replied:
This is the recommended definition as per the [Resource Manager spec](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-resources/resources/2016-09-01/swagger/resources.json#L47). This should be applied consistently across all the RPs.

```
{
            "name": "resourceGroupName",
            "in": "path",
            "required": true,
            "type": "string",
            "description": "The name of the resource group with the deployment to delete. The name is case insensitive.",
            "pattern": "^[-\\w\\._\\(\\)]+$",
            "minLength": 1,
            "maxLength": 90
          },
```

fearthecowboy commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+            "b",
+            "q",
+            "t",
+            "f"
```
Ouch! >.<

Is there any way these could be actual values? There's no way to tell what these mean.

(FYI; It's a current limitation in swagger that there isn't any means to document the meaning of enum values. -- we're investigating ways to improve that)

fearthecowboy commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+            "o"
+          ],
+          "x-ms-client-name": "ResourceTypes",
+          "description": "Sets the signed resource types that are accessible with the account SAS."
```
>Sets [](start = 26, length = 4)

Omit the word 'sets' here.

fearthecowboy commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+            "f"
+          ],
+          "x-ms-client-name": "Services",
+          "description": "Sets the signed services accessible with the account SAS."
```
>Sets [](start = 26, length = 4)

Omit the word 'sets' here

fearthecowboy commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+            "minLength": 3
+          },
+          {
+            "name": "regenerateKey",
```
Another good opportunity to use x-ms-client-flatten. Maybe take a run thru where you have body parameters, and if they're 1-3 properties, see if you can use x-ms-client-flatten. Makes the APIs so much easier to use for customers. :D

robinsh commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+        "description": "Checks that the storage account name is valid and is not already in use.",
+        "parameters": [
+          {
+            "name": "accountName",
```
It has been called accountName for a very long time, and it is account name to the customer, account name on the portal, and is a value selected and set by the customer. I don't understand why we would want to change it after all this time. Or am I misunderstanding your comment?

robinsh commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -529,7 +529,8 @@
         "type"
-      ]
+      ],
+      "description": "The parameters used to check the availabity of storage account name."
```
Can you add a "the" in there, so it says "The parameters used to check the availability of the storage account name." ?

robinsh commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -672,7 +673,8 @@
         }
-      }
+      },
+      "description": "The parameters used to create storage account."
```
Can you add a "the" in there, so it is "The parameters used to create the storage account." ?

robinsh commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -837,7 +839,8 @@
         }
-      }
+      },
+      "description": "Properties of storage account."
```
Can you change this to "Properties of the storage account." ?

robinsh commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -933,7 +936,8 @@
         "keyName"
-      ]
+      ],
+      "description": "The parameters used to regenerate storage account key."
```
Can you please add "the" before "storage account key." ?

robinsh commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -1074,7 +1079,8 @@
       },
-      "x-ms-azure-resource": true
+      "x-ms-azure-resource": true,
+      "description": "Describes storage resource."
```
"Describes a storage resource."  ?

amarzavery commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "The name of the resource group within the user's subscription."
```
please add the regex pattern and the min/max constraint everywhere for ResourceGroup as explained in the previous comments.
ResourceGroupName can be defined in the global parameters once, with the extension `"x-ms-parameter-location": "method"` and referenced everywhere. This reduces duplication.

JasonYang-MSFT replied:
"x-ms-parameter-location" is supported from autorest 0.17, right? If so, will upgrade autorest to 0.17 for .net sdk accordingly.

amarzavery replied:
AutoRest.1.0.0-Nightly20161122 is the latest nightly build. Pleas use that. That gives you all the new and good features.

amarzavery commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+      "properties": {
+        "canonicalizedResource": {
+          "type": "string",
+          "x-ms-client-name": "CanonicalizedResource",
```
How is the client-name different from the property name except it being PascalCase. This property will be pascal cased any ways in C# as the model property name and the serialized name would remain as is.  In javascript no matter what the model property name would remain camelCase. So, I don't see the need over here. Please remove this extension.

amarzavery commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+            "s"
+          ],
+          "x-ms-client-name": "Resource",
+          "description": "Sets the signed services accessible with the service SAS."
```
General comment: Can you please describe these single letter enum values in their descriptions? That would be of great help to the customer.

amarzavery commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1327 @@
+        "signedIdentifier": {
+          "type": "string",
+          "x-ms-client-name": "Identifier",
+          "description": "Sets a unique value up to 64 characters in length that correlates to an access policy specified for the container, queue, or table."
```
> description says: upto 64 characters in length.
please specify the `"maxLength": 64` constraint.

JasonYang-MSFT committed:
[Storage] updated pr according to the comments.

1. Changed resourceGroupName to global and add constraints.
2. Changed tile to fit for website.
3. Removed "text/json".
4. Added more descriptions for enum type values.
5. Fixed other document issues.

JasonYang-MSFT commented:
Currently, comments are resolved. Please take a took and comment back if any further comments.

amarzavery commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1301 @@
+          "x-ms-enum": {
+            "name": "AccountStatus",
+            "modelAsString": false
+          }
```
@JasonYang-MSFT - We have built a validation tool that validates an example request body and the response against the schema defined in the swagger spec. The tool fails on an error. 
The service returns value for "statusOfPrimary" as "available" instead of PascalCase "Available". This causes our validation to fail. I think the service should normalize the value to PascalCase while sending the value back to the customer. It can be relaxed while accepting the value. This will ensure that applications do not break on deserialization.

Sample output of the validation tool
```
> Operation: StorageAccounts_GetProperties

                > x-ms-examples

                        > Scenario: GetProperties
                                > RESPONSE_BODY_VALIDATION_ERROR: Found errors in validating the response with statusCode "200" for example "GetProperties" in operation "StorageAccounts_GetProperties".
                                > InnerErrors:
                                  1. code: ENUM_MISMATCH
                                     message: No enum match for: available
                                     path: properties,statusOfPrimary
                                     description: Gets the status indicating whether the primary location of the storage account is available or unavailable.
```

amarzavery replied:
[This is the format](https://github.com/Azure/openapi-validation-tools/blob/master/arm-storage/2016-01-01/swagger/storage.json#L65) in which we are providing samples for an operation. 
[This is the actual sample](https://github.com/Azure/openapi-validation-tools/blob/master/arm-storage/2016-01-01/examples/storageAccountCreate.json) in a separate json file.
It captures example/sample values for everything that the user needs to make a request (path/query/body parameters). It also has a sample value for the response body and any important headers (for long running operations). This helps us ensure that the actual wire format and the swagger spec are in sync. The tool validates all the sample values against the spec.

Moving forward we would want all the service teams to provide samples as a part of their PR. It would be great if you can use the sample that we already have for the 2016-01-01 api-version of storage spec and add samples for the 2 new operations added in this api-version. 

the doc team is going to take all these samples and lighten them in the docs.microsoft.com website in the next sprint.

JasonYang-MSFT replied:
Yes, server side actually returned as available or unavailable. And this behavior lasts at least one year. Changing to PascalCase would cause a breaking on rest api level. 
I think it's better to sync on sdk side to change Available to available. 
For future design, we can try to return PascalCase values.

JasonYang-MSFT replied:
@amarzavery 
Seems unable to reply in your newly reply, so updated here, and replied inline.

1. This is a new api-version (2016-05-01) so how will you break customers?
[Jason] this is an incremental update of 2016-01-01. For the part we are discussing, they are shared with version 2016-01-01 for now. 

2. Fixing this in all the sdks is not the right solution.
[Jason]Yes. Fixing in swagger will reflect the changes to other sdks.

3. The storage spec, the way it is right now, is currently broken. Hardening the response from the server side is not a breaking change. In fact the response from the service will conform to the enum values specified in the swagger spec.

4. If you cannot fix the service then please fix the swagger spec. Please add all the variations in the enum that the server sends.
[Jason]Yes, fix should be in swagger spec. But not quite understand "the variations in the enum that the server sends"? Like following part?

        "statusOfPrimary": {
          "readOnly": true,
          "type": "string",
          "description": "Gets the status indicating whether the primary location of the storage account is available or unavailable.",
          "enum": [
            "available",
            "unavailable"
          ],
          "x-ms-enum": {
            "name": "AccountStatus",
            "modelAsString": false
          }
        },

Currently REST API docs on the doc website (docs.microsoft.com) stand incorrect because the swagger spec is incorrect. Please fix this ASAP.
[Jason] docs didn't reflect the enum part. And the description part is correct for "statusOfSecondary". For enum value in "StorageAccountKey", the value should be "Full" or "Read", will confirm that latter and I think it's better to fix in another PR.

amarzavery commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -0,0 +1,1301 @@
+          "x-ms-enum": {
+            "name": "KeyPermission",
+            "modelAsString": false
+          }
```
Same is the case with permissions.
```
> Operation: StorageAccounts_ListKeys

                > x-ms-examples

                        > Scenario: ListKeys
                                > RESPONSE_BODY_VALIDATION_ERROR: Found errors in validating the response with statusCode "200" for example "ListKeys" in operation "StorageAccounts_ListKeys".
                                > InnerErrors:
                                  1. code: ENUM_MISMATCH
                                     message: No enum match for: Full
                                     path: keys,1,permissions
                                     description: Permissions for the key -- read-only or full permissions.
                                  2. code: ENUM_MISMATCH
                                     message: No enum match for: Full
                                     path: keys,0,permissions
                                     description: Permissions for the key -- read-only or full permissions.
```

amarzavery commented:
@JasonYang-MSFT -


that is not correct.

1. This is a new api-version (2016-05-01) so how will you break customers?

2. Fixing this in all the sdks is not the right solution.

3. The storage spec, the way it is right now, is currently broken. Hardening the response from the server side is not a breaking change. Infact the response from the service will conform to the enum values specified in the swagger spec.

4. If you cannot fix the service then please fix the swagger spec. Please add all the variations in the enum that the server sends.


Currently REST API docs on the doc website (docs.microsoft.com) stand incorrect because the swagger spec is incorrect. Please fix this ASAP.

________________________________
From: Jason Yang <notifications@github.com>
Sent: Wednesday, November 23, 2016 11:40:50 PM
To: Azure/azure-rest-api-specs
Cc: Amar Zavery; Mention
Subject: Re: [Azure/azure-rest-api-specs] [Storage] ListAccountSas and ListServiceSas (#742)


@JasonYang-MSFT commented on this pull request.

________________________________

In arm-storage/2016-05-01/swagger/storage.json<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F742&data=02%7C01%7Camzavery%40microsoft.com%7Ca008d25020874bafd80f08d4143d36f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636155700525623785&sdata=tnpEOXSBqhYG9wbjFOKITWgjE9EXvDrP2XNjYaxYgC8%3D&reserved=0>:

> +          "readOnly": true,
+          "type": "string",
+          "description": "Gets the location of the primary data center for the storage account."
+        },
+        "statusOfPrimary": {
+          "readOnly": true,
+          "type": "string",
+          "description": "Gets the status indicating whether the primary location of the storage account is available or unavailable.",
+          "enum": [
+            "Available",
+            "Unavailable"
+          ],
+          "x-ms-enum": {
+            "name": "AccountStatus",
+            "modelAsString": false
+          }


Yes, server side actually returned as available or unavailable. And this behavior lasts at least one year. Changing to PascalCase would cause a breaking on rest api level.
I think it's better to sync on sdk side to change Available to available.
For future design, we can try to return PascalCase values.

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-rest-api-specs%2Fpull%2F742&data=02%7C01%7Camzavery%40microsoft.com%7Ca008d25020874bafd80f08d4143d36f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636155700525623785&sdata=tnpEOXSBqhYG9wbjFOKITWgjE9EXvDrP2XNjYaxYgC8%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAET9BzFpqqn6wWOmmyvCIn6uoJEeWbIsks5rBT-CgaJpZM4K2S_L&data=02%7C01%7Camzavery%40microsoft.com%7Ca008d25020874bafd80f08d4143d36f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636155700525633787&sdata=F3zvE3%2F0T%2BobAMRZ%2BnRifUUK2dVMyzHibz2pprThlIc%3D&reserved=0>.

JasonYang-MSFT committed:
[Storage] update typo.

JasonYang-MSFT commented:
Updated to align enum values exactly same to response body.

amarzavery commented on arm-storage/2016-05-01/swagger/storage.json:
```diff
@@ -854,8 +854,8 @@
-            "READ",
-            "FULL"
+            "Read",
+            "Full"
```
@veronicagg, @johanste - With these changes the data validation error for enum mismatch will go away. @jhendrixMSFT, @markcowl, @lmazuel - Heads up. This is a good breaking change. It change the valid enum values in the new version of storage management sdk in all the languages. Since all the properties are readOnly, the effect would be minimal. I assume there shouldnt be any breaking change in Powershell and CLI (api wise).

amarzavery commented:
@fearthecowboy - All the changes look good to me. Please feel free to merge the PR if you are satisfied with the changes as well.

JasonYang-MSFT commented:
Hi @fearthecowboy 
Any other new comments?

amarzavery commented:
I think he must be busy. I am merging this PR. LGTM.

AutorestCI commented:
No modification for Ruby

AutorestCI commented:
No modification for NodeJS

AutorestCI commented:
No modification for Python

fearthecowboy commented:
_sorry on the delay-- it's been a busy few days_ 

Yeah, I had no additional feedback.
